### PR TITLE
Outbound reminder sample for Java

### PR DESCRIPTION
**Pull request by ravithanneeru**
## Purpose

* ...

## Does this introduce a breaking change?

```
[ ] Yes
[ ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: OutboundCallReminder/.project [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/.project:1]: Can we have this project as a pure maven project, not eclipse project and not check-in any eclipse ide specific file? I see the other samples are created as mvn projects only as well.

File: OutboundCallReminder/.classpath [hunk 2]
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/.classpath:1]: Can we have this project as a pure maven project, not eclipse project and not check-in any eclipse ide specific file? I see the other samples are created as mvn projects only as well.

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/App.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            ConfigurationManager configurationManager = ConfigurationManager.GetInstance();
+            String ngrokPath = configurationManager.GetAppSettings("NgrokExePath");
+
+            if (ngrokPath.isEmpty()) {
+                Logger.LogMessage(Logger.MessageType.INFORMATION, "Ngrok path not provided");
+                return null;
+            }
+
+            Logger.LogMessage(Logger.MessageType.INFORMATION,"Starting Ngrok");
+            ngrokService = new NgrokService(ngrokPath, null);
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/App.java:68]: Looks like we are passing null for ngrok authToken. Don't we need to pass a valid value, possibly read from config?

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/Ngrok/NgrokService.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    private void CreateNgrokProcess(String ngrokPath, String authToken) {
+        try {
+            String authTokenArgs = "";
+            if (authToken != null && !authToken.isEmpty()) {
+                authTokenArgs = "--authtoken " + authToken;
+            }
+
+            String openCmd = "cmd /c start cmd.exe /k ";
+            String fileName = ngrokPath + "ngrok.exe";
+            String arguments = " http http://localhost:9007/ -host-header=/localhost:9007/" + authTokenArgs;
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/Ngrok/NgrokService.java:68]: Don't we need a space before the 'authTokenArgs'. For example:
String arguments = " http http://localhost:9007/ -host-header=/localhost:9007/" + " " +authTokenArgs;

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/App.java (hunk 3 above)
Comment (ravithanneeru) [reply OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/App.java:68]: Not required, at least in our samples case. As per ngrok docs, authtoken require for some advanced features.

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/Ngrok/NgrokService.java (hunk 4 above)
Comment (ravithanneeru) [reply OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/Ngrok/NgrokService.java:68]: Yes, correct, we need to add space. Since we are not passing authtoken, so authTokenArgs value is empty, so not causing any issue at the moment. Will add space before.

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+public class EventDispatcher {
+    private static EventDispatcher instance = null;
+    private Hashtable<String, NotificationCallback> NotificationCallback;
+
+    EventDispatcher() {
+        NotificationCallback = new Hashtable<String, NotificationCallback>();
+    }
+
+    /// <summary>
+    /// Get istace of EventDispatcher
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:17]: nit: typo

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
+package com.communication.outboundcallreminder.EventHandler;
+
+import com.azure.communication.callingserver.models.events.*;
+import com.azure.core.models.CloudEvent;
+import com.azure.core.util.BinaryData;
+import java.util.*;
+
+public class EventDispatcher {
+    private static EventDispatcher instance = null;
+    private Hashtable<String, NotificationCallback> NotificationCallback;
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:10]: can we name this to variable as 'notificationCallbacks' or similar

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    /// Get istace of EventDispatcher
+    /// </summary>
+    public static EventDispatcher GetInstance() {
+        if (instance == null) {
+            instance = new EventDispatcher();
+        }
+        return instance;
+    }
+
+    public Boolean Subscribe(String eventType, String eventKey, NotificationCallback notificationCallback) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:26]: Lets rename method name according to java naming guideline. 
https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

It should be 'subscribe()'

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        String eventId = BuildEventKey(eventType, eventKey);
+        synchronized (this) {
+            return (NotificationCallback.put(eventId, notificationCallback) == null);
+        }
+    }
+
+    public void Unsubscribe(String eventType, String eventKey) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:33]: please fix method name

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        String eventId = BuildEventKey(eventType, eventKey);
+        synchronized (this) {
+            NotificationCallback.remove(eventId);
+        }
+    }
+
+    public String BuildEventKey(String eventType, String eventKey) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:40]: please fix method name

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        return (eventType + "-" + eventKey);
+    }
+
+    public void ProcessNotification(String request) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:44]: please fix method name

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                if (notificationCallback != null) {
+                    new Thread(() -> {
+                        notificationCallback.Callback(callEvent);
+                    }).start();
+                }
+            }
+        }
+    }
+
+    public String GetEventKey(CallingServerEventBase callEventBase) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:59]: please fix method name

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+            return BuildEventKey(CallingServerEventType.PLAY_AUDIO_RESULT_EVENT.toString(), operationContext);
+        } else if (callEventBase.getClass() == AddParticipantResultEvent.class) {
+            String operationContext = ((AddParticipantResultEvent) callEventBase).getOperationContext();
+            return BuildEventKey(CallingServerEventType.ADD_PARTICIPANT_RESULT_EVENT.toString(), operationContext);
+        }
+        
+        return null;
+    }
+
+    public CallingServerEventBase ExtractEvent(String content) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:77]: please fix method name

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java (hunk 7 above)
Comment (arifsaikat-microsoft) [reply OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/EventHandler/EventDispatcher.java:26]: please fix method names in other parts of the codebase

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/Controllers/OutboundCallController.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import java.io.FileInputStream;
+import com.communication.outboundcallreminder.Logger;
+import com.communication.outboundcallreminder.EventHandler.EventAuthHandler;
+import com.communication.outboundcallreminder.EventHandler.EventDispatcher;
+
+@RestController
+public class OutboundCallController {
+
+	@RequestMapping("/api/outboundcall/callback")
+	public static String OnIncomingRequestAsync(@RequestBody(required = false) String data,
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/Controllers/OutboundCallController.java:18]: please rename method name

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/CallConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
+package com.communication.outboundcallreminder;
+
+import com.communication.outboundcallreminder.EventHandler.EventAuthHandler;
+
+public class CallConfiguration {
+    public String ConnectionString;
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/CallConfiguration.java:6]: please fix variable names

File: PhoneNumbers/pom.xml [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+  <dependencies>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.11</version>
+      <scope>test</scope>
+    </dependency>
+	<dependency>
+		<groupId>com.azure</groupId>
```
Comment (arifsaikat-microsoft) [inline PhoneNumbers/pom.xml:29]: please fix alignment

File: PhoneNumbers/pom.xml [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+		<artifactId>azure-communication-phonenumbers</artifactId>
+		<version>1.0.0</version>
+	</dependency>
+	<dependency>
+		<groupId>com.azure</groupId>
```
Comment (arifsaikat-microsoft) [inline PhoneNumbers/pom.xml:34]: please fix alignment

File: PhoneNumbers/pom.xml [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+		<artifactId>azure-communication-common</artifactId>
+		<version>1.0.0</version>
+	</dependency>
+	<dependency>
+		<groupId>com.azure</groupId>
```
Comment (arifsaikat-microsoft) [inline PhoneNumbers/pom.xml:39]: please fix alignment

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 18]
```diff
@@ -0,0 +1,362 @@
...
+            // Wait for the call to terminate
+            callTerminatedTask.get();
+        } catch (Exception ex) {
+            Logger.logMessage(Logger.MessageType.ERROR, "Call ended unexpectedly, reason -- > " + ex.getMessage());
+        }
+    }
+
+    private void createCallAsync(String targetPhoneNumber) {
+        try {
+            // Preparting request data
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:80]: typo

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 19]
```diff
@@ -0,0 +1,362 @@
...
+            EventDispatcher.getInstance().unsubscribe(CallingServerEventType.TONE_RECEIVED_EVENT.toString(), callLegId);
+            // cancel playing audio
+            cancelMediaProcessing(callLegId);
+        });
+        // Subscribe to event
+        EventDispatcher.getInstance().subscribe(CallingServerEventType.TONE_RECEIVED_EVENT.toString(), callLegId,
+                dtmfReceivedEvent);
+    }
+
+    private void cancelMediaProcessing(String callLegId) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:160]: is the param callLegId used in the method?

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 20]
```diff
@@ -0,0 +1,362 @@
...
+        String operationContext = UUID.randomUUID().toString();
+        Response<CancelAllMediaOperationsResult> cancelmediaresponse = this.callConnection.cancelAllMediaOperationsWithResponse(operationContext, null);
+        CancelAllMediaOperationsResult response = cancelmediaresponse.getValue();
+
+        Logger.logMessage(Logger.MessageType.INFORMATION, "cancelAllMediaOperationsWithResponse -- > " + getResponse(cancelmediaresponse) + 
+        ", Id: " + response.getOperationId() + ", OperationContext: " + response.getOperationContext() + ", OperationStatus: " +
+        response.getStatus().toString());
+    }
+
+    private void playAudioAsync(String callLegId) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:177]: is the param callLegId used in the method?

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 21]
```diff
@@ -0,0 +1,362 @@
...
+        } catch (Exception ex) {
+            if (playAudioCompletedTask.isCancelled()) {
+                Logger.logMessage(Logger.MessageType.INFORMATION, "Play audio operation cancelled");
+            } else {
+                Logger.logMessage(Logger.MessageType.INFORMATION, "Failure occured while playing audio on the call. Exception: " + ex.getMessage());
+            }
+        }
+    }
+
+    private void hangupAsync(String callLegId) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:234]: is the param callLegId used in the method?

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 22]
```diff
@@ -0,0 +1,362 @@
...
+            if (addParticipantResult) {
+                return;
+            } else {
+                Logger.logMessage(Logger.MessageType.INFORMATION, "Retry add participant attempt -- > " + retryAttemptCount + " has failed");
+                retryAttemptCount++;
+            }
+        }
+    }
+
+    private Boolean addParticipant(String callLegId, String addedParticipant) {
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:280]: is the param callLegId used in the method?

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+    private CommunicationIdentifierKind getIdentifierKind(String participantnumber) {
+        // checks the identity type returns as string
+        return ((Pattern.matches(userIdentityRegex, participantnumber)) ? CommunicationIdentifierKind.UserIdentity
+                : (Pattern.matches(phoneIdentityRegex, participantnumber)) ? CommunicationIdentifierKind.PhoneIdentity
+                        : CommunicationIdentifierKind.UnknownIdentity);
+    }
+
+    public String getResponse(Response<?> response)
+    {
+        String responseString = null;
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:350]: nit: may be use a StringBuilder class here, that will be more efficient for concatenating

File: PhoneNumbers/pom.xml [hunk 24]
```diff
@@ -0,0 +1,90 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (arifsaikat-microsoft) [inline PhoneNumbers/pom.xml:1]: Is this file related to the two samples?

File: OutboundCallReminder/pom.xml [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+
+  <dependencies>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.11</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+			<groupId>org.springframework.boot</groupId>
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/pom.xml:36]: fix alignment

File: OutboundCallReminder/pom.xml [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+			<artifactId>spring-boot-starter-web</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
```
Comment (arifsaikat-microsoft) [inline OutboundCallReminder/pom.xml:40]: fix alignment

File: ServerRecording/pom.xml [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+			<artifactId>spring-boot-starter-test</artifactId>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.httpcomponents</groupId>
+			<artifactId>httpclient</artifactId>
+			<version>4.5.13</version>
+		</dependency>
+		<dependency>
+			<groupId>com.googlecode.json-simple</groupId>
```
Comment (arifsaikat-microsoft) [inline ServerRecording/pom.xml:46]: can we remove this dependency, same as the project?

File: PhoneNumbers/pom.xml (hunk 24 above)
Comment (ravithanneeru) [reply PhoneNumbers/pom.xml:1]: No, it's a different sample pending from some time. Missed to merge.

File: OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java [hunk 28]
```diff
@@ -356,6 +356,15 @@ private void registerToAddParticipantsResultEvent(String operationContext) {
             OperationStatus operationStatus = addParticipantsUpdatedEvent.getStatus();
             if (operationStatus.equals(OperationStatus.COMPLETED)) {
                 Logger.logMessage(Logger.MessageType.INFORMATION, "Add participant status -- > " + operationStatus);
+                
+                //Remove this sleep logic once the call disconnection issue when two PSTN participants present in the call and app exited
```
Comment (satyakonmsft) [inline OutboundCallReminder/src/main/java/com/communication/outboundcallreminder/OutboundCallReminder.java:360]: let us remove this comment. our documentation's known issues will already address this issue
